libraries/AP_IBus_Telem/AP_IBus_Telem.cpp

```diff
@@ -0,0 +1,193 @@
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*
+   IBus telemetry should always run at 115200 baud rate.
+   It uses a half duplex protocol, so you need only to wire the TX pin to the telemetry port of the receiver
+   You can configure via your preferred GCS the serial port to send IBus telemetry data
+   You  need to initialize this object and then add sensors and set sensor measurements (about every 500ms - 1sec)
+*/
+
+#include <AP_IBus_Telem/AP_IBus_Telem.h>
+#include <AP_BattMonitor/AP_BattMonitor.h>
+#include <AP_Vehicle/AP_Vehicle.h>
+#include <AP_Arming/AP_Arming.h>
+
+#if HAL_IBUS_TELEM_ENABLED
+
+void AP_IBus_Telem::init()
+{
+    const AP_SerialManager &serial_manager = AP::serialmanager();
+    if ((_port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_IBUS_Telem, 0))) {
+        _port->set_options(_port->OPTION_HDPLEX);
+        _port->begin(AP_SERIALMANAGER_IBUS_TELEM_BAUD);
+        _initialized = true;
+    }
+}
+
+void AP_IBus_Telem::populate_checksum(uint8_t *packet, uint16_t size)
+{
+    uint16_t checksum = 0xFFFF;
+
+    for (int i=0; i<size-2; i++) {
+        checksum -= *packet;
+        packet++;
+    }
+
+    *packet = checksum & 0x0ff;
+    packet++;
+    *packet = checksum >> 8;
+}
+
+void AP_IBus_Telem::loop(void)
+{
+    //no need to run the loop if not initialized or no sensors added
+    if (!_initialized) {
+        return;
+    }
+
+    uint32_t now = AP_HAL::millis();
+    if (now - _last_send_measurements_time > UPDATE_MEASUREMENTS_TIMING) {
+        setSensorMeasurements();
+    }
+
+    // only process data already in our UART receive buffer
+    while (_port->available() > 0) {
+        // only consider a new data package if we have not heard anything for >3ms
+        if (now - _last >= PROTOCOL_TIMEGAP) {
+            _state = GET_LENGTH;
+        }
+        _last = now;
+
+        int16_t c = _port->read();
+        if (c < 0) {
+            return;
+        }
+
+        switch (_state) {
+        case GET_LENGTH:
+            if (c <= PROTOCOL_MAX_LENGTH && c > PROTOCOL_OVERHEAD) {
+                _ptr = 0;
+                _len = c - PROTOCOL_OVERHEAD;
+                _chksum = 0xFFFF - c;
+                _state = GET_DATA;
+            } else {
+                _state = DISCARD;
+            }
+            break;
+
+        case GET_DATA:
+            _buffer[_ptr++] = c;
+            _chksum -= c;
+            if (_ptr == _len) {
```

Comment:
```suggestion
            if (_ptr >= _len) {
```

(should never happen, but its defensive programming which will minimise a buffer overrun)